File: web-static/js/mesh.js
// Склейка одноцветных клеток в масштабированные блоки. Порт core/mesh.py.
//
// Scrap Mechanic умеет хранить блок растянутым (поле "bounds"), и в чертежах
// самой игры встречаются блоки с bounds до 892 — значит «одна деталь на
// пиксель» это чистая потеря. Здесь жадное разбиение на прямоугольники:
// из каждой свободной клетки тянем максимально широкую полосу одного
// материала, затем опускаем её вниз, пока строки под ней совпадают.
//
// Порядок обхода и содержимое прямоугольников обязаны совпадать с
// питоновской версией — это проверяют эталонные векторы.

export const MAX_BOUND_DEFAULT = 255;

/** Упаковать RGB в один ключ: 0xRRGGBB. */
export function colorKeys(rgb, n) {
  const keys = new Int32Array(n);
  for (let i = 0; i < n; i++) {
    keys[i] = (rgb[i * 3] << 16) | (rgb[i * 3 + 1] << 8) | rgb[i * 3 + 2];
  }
  return keys;
}

/** Без слияния: каждая клетка — отдельный блок 1x1x1. */
export function splitRects(keys, mask, w, h) {
  const rects = [];
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const i = y * w + x;
      if (mask && !mask[i]) continue;
      rects.push([x, y, 1, 1, keys[i]]);
    }
  }
  return rects;
}

/**
 * Жадное разбиение сетки на одинаковые по материалу прямоугольники.
 * keys — Int32Array w*h, mask — Uint8Array w*h (или null, если дыр нет).
 * Возвращает массив [x, y, ширина, высота, ключ].
 */
export function mergeRects(keys, mask, w, h, maxBound = MAX_BOUND_DEFAULT) {
  maxBound = Math.max(1, maxBound | 0);
  if (maxBound === 1) return splitRects(keys, mask, w, h);

  const at = (x, y) => (mask ? mask[y * w + x] : 1);

  // csd[y][x] — сколько подряд идущих столбцов до x совпали со строкой снизу.
  // Префиксные суммы дают проверку «вся полоса совпадает» за две операции.
  let csd = null;
  if (h > 1) {
    csd = new Int32Array((h - 1) * (w + 1));
    for (let y = 0; y < h - 1; y++) {
      const base = y * (w + 1);
      let acc = 0;
      for (let x = 0; x < w; x++) {
        const i = y * w + x;
        const same = at(x, y) && at(x, y + 1) && keys[i] === keys[i + w] ? 1 : 0;
        acc += same;
        csd[base + x + 1] = acc;
      }
    }
  }

  const rects = [];
  // Клетка занята <=> попала в уже созданный прямоугольник. Так как они
  // растут только вниз и мы идём сверху вниз, свободная клетка гарантирует,
  // что всё под ней в этом столбце тоже свободно.
  const takenUntil = new Int32Array(w);

  for (let y = 0; y < h; y++) {
    const rowOff = y * w;

    // границы одноцветных отрезков в строке — считаем один раз на строку
    let runStart = 0;
    while (runStart < w) {
      let runEnd = runStart + 1;
      while (runEnd < w
        && keys[rowOff + runEnd] === keys[rowOff + runEnd - 1]
        && at(runEnd, y) === at(runEnd - 1, y)) runEnd++;

      if (at(runStart, y)) {
        let x = runStart;
        while (x < runEnd) {
          if (takenUntil[x] > y) { x++; continue; }

          const limit = Math.min(runEnd, x + maxBound);
          let width = 1;
          while (x + width < limit && takenUntil[x + width] <= y) width++;

          let height = 1;
          if (csd) {
            const maxH = Math.min(h - y, maxBound);
            while (height < maxH) {
              const yy = y + height - 1;
              const base = yy * (w + 1);
              if (csd[base + x + width] - csd[base + x] !== width) break;
              height++;
            }
          }

          rects.push([x, y, width, height, keys[rowOff + x]]);
          for (let k = 0; k < width; k++) takenUntil[x + k] = y + height;
          x += width;
        }
      }
      runStart = runEnd;
    }
  }
  return rects;
}

/** Сводка для интерфейса. */
export function stats(rects) {
  if (!rects.length) return { parts: 0, cells: 0, colors: 0, biggest: 0, mergedRatio: 0 };
  let cells = 0, biggest = 0;
  const colors = new Set();
  for (const [, , w, h, key] of rects) {
    const area = w * h;
    cells += area;
    if (area > biggest) biggest = area;
    colors.add(key);
  }
  return {
    parts: rects.length,
    cells,
    colors: colors.size,
    biggest,
    mergedRatio: cells ? +(1 - rects.length / cells).toFixed(4) : 0,
  };
}

File: core/mesh.py
"""Слияние одноцветных пикселей в масштабированные блоки.

Scrap Mechanic умеет хранить блок растянутым: поле "bounds" в чертеже.
В собственных чертежах игры встречаются блоки с bounds до 892 — то есть
ограничения практически нет, и «одна деталь на пиксель» это чистая потеря.

Здесь — жадное разбиение на прямоугольники (greedy meshing): идём слева
направо и сверху вниз, из каждой свободной клетки тянем максимально
широкую полосу одного цвета, затем опускаем её вниз, пока строки под ней
совпадают. Картинка получается пиксель-в-пиксель та же, а деталей
становится в разы меньше.

Ни один из существующих конвертеров такого не делает: два веб-конвертера
пишут ровно 1x1x1 на пиксель, а единственный, который что-то склеивает
(SMBIC), склеивает только горизонтальные отрезки внутри одной строки.
"""

from __future__ import annotations

import numpy as np

# Прямоугольник: (x, y, ширина, высота, цвет как 0xRRGGBB)
Rect = tuple[int, int, int, int, int]

MAX_BOUND_DEFAULT = 255


def color_keys(rgb: np.ndarray) -> np.ndarray:
    """(H, W, 3) uint8 -> (H, W) uint32, один цвет = одно число."""
    a = rgb.astype(np.uint32)
    return (a[..., 0] << 16) | (a[..., 1] << 8) | a[..., 2]


def as_keys(grid: np.ndarray) -> np.ndarray:
    """Принять либо готовые ключи (H, W), либо картинку (H, W, 3)."""
    arr = np.asarray(grid)
    return color_keys(arr) if arr.ndim == 3 else arr


def split_rects(keys: np.ndarray, mask: np.ndarray) -> list[Rect]:
    """Без слияния: каждая клетка — отдельный блок 1x1x1."""
    keys = as_keys(keys)
    ys, xs = np.nonzero(mask)
    vals = keys[ys, xs]
    return [(int(x), int(y), 1, 1, int(c)) for x, y, c in zip(xs, ys, vals)]


def merge_rects(keys: np.ndarray, mask: np.ndarray, max_bound: int = MAX_BOUND_DEFAULT) -> list[Rect]:
    """Жадное разбиение сетки на одинаковые по материалу прямоугольники.

    keys — целочисленный «что за блок и какого цвета» для каждой клетки.
    Соседи склеиваются только если ключ совпал полностью, поэтому режим с
    разными блоками работает тем же кодом, что и одноцветный.
    """
    key = np.ascontiguousarray(as_keys(keys), dtype=np.int64)
    h, w = key.shape
    max_bound = max(1, int(max_bound))
    if max_bound == 1:
        return split_rects(key, mask)

    # same_down[y, x] — клетка снизу того же цвета и тоже не пустая.
    if h > 1:
        same_down = mask[:-1] & mask[1:] & (key[:-1] == key[1:])
        # префиксные суммы -> проверка «вся полоса совпадает» за две операции
        csd = np.zeros((h - 1, w + 1), dtype=np.int32)
        np.cumsum(same_down, axis=1, out=csd[:, 1:], dtype=np.int32)
    else:
        csd = None

    rects: list[Rect] = []
    # Клетка занята <=> она попала в уже созданный прямоугольник. Так как
    # прямоугольники растут только вниз и мы идём сверху вниз, свободная
    # клетка гарантирует, что всё под ней в этом столбце тоже свободно.
    taken_until = np.zeros(w, dtype=np.int32)  # первая строка, с которой столбец снова свободен

    for y in range(h):
        row = key[y]
        rmask = mask[y]
        # границы одноцветных отрезков в строке — считаем один раз на строку
        change = np.empty(w, dtype=bool)
        change[0] = True
        np.not_equal(row[1:], row[:-1], out=change[1:])
        change[1:] |= rmask[1:] != rmask[:-1]
        starts = np.flatnonzero(change)
        ends = np.empty_like(starts)
        ends[:-1] = starts[1:]
        ends[-1] = w

        for run_start, run_end in zip(starts, ends):
            if not rmask[run_start]:
                continue
            x = int(run_start)
            stop = int(run_end)
            while x < stop:
                if taken_until[x] > y:
                    x += 1
                    continue
                seg_end = stop
                # столбцы правее могут быть заняты «свисающим» прямоугольником
                limit = min(seg_end, x + max_bound)
                width = 1
                while x + width < limit and taken_until[x + width] <= y:
                    width += 1

                height = 1
                if csd is not None:
                    max_h = min(h - y, max_bound)
                    while height < max_h:
                        yy = y + height - 1
                        if csd[yy, x + width] - csd[yy, x] != width:
                            break
                        height += 1

                rects.append((x, y, width, height, int(row[x])))
                taken_until[x:x + width] = y + height
                x += width

    return rects


def stats(rects: list[Rect]) -> dict:
    """Сводка для интерфейса."""
    if not rects:
        return {"parts": 0, "cells": 0, "colors": 0, "biggest": 0, "merged_ratio": 0.0}
    cells = sum(r[2] * r[3] for r in rects)
    colors = len({r[4] for r in rects})
    biggest = max(r[2] * r[3] for r in rects)
    return {
        "parts": len(rects),
        "cells": cells,
        "colors": colors,
        "biggest": biggest,
        "merged_ratio": round(1 - len(rects) / cells, 4) if cells else 0.0,
    }
